infrastructure/merklemountainrange/src/merklemountainrange.rs

```diff
@@ -338,10 +359,45 @@ where
         }
         if (new_index <= self.get_last_added_index()) && (height >= 0) {
             // is this a valid peak which needs to be bagged
-            peaks.push(self.mmr[new_index].hash.clone());
+            peaks.push(self.mmr[new_index].clone());
             self.find_bagging_indexes(height, new_index, peaks); // lets go look for more peaks
         }
     }
+
+    /// Mark an object as pruned, if the MMR can remove this safely it will
+    pub fn prune_object_hash(&mut self, hash: ObjectHash) -> Result<(), MerkleMountainRangeError> {
+        let object = self.data.get_mut(&hash);
+        if object.is_none() {
+            return Err(MerkleMountainRangeError::ObjectNotFound);
+        };
+        let object = object.unwrap();
+        object.pruned = true;
+        let sibling_index = if is_node_right(object.vec_index) {
+            object.vec_index - 1
+        } else {
+            object.vec_index + 1
+        };
+        let sibling = self.data.get_mut(&self.mmr[sibling_index]);
+        if sibling.is_none() {
+            return Ok(());
+        };
+        let sibling = sibling.unwrap();
+        if sibling.pruned == true {
+            self.data.remove(&self.mmr[sibling_index]);
+            self.data.remove(&hash);
+        }
+        Ok(())
+    }
+
+    /// Mark an object as pruned, if the MMR can remove this safely it will
+    pub fn prune_index(&mut self, object_index: usize) -> Result<(), MerkleMountainRangeError> {
+        if object_index > self.data.len() {
+            return Err(MerkleMountainRangeError::IndexOutOfBounds);
+        }
+        let index = get_object_index(object_index);
+        let hash = self.mmr[index].clone();
```

Comment:
No need to clone if you change the method signature above.

Reply:
Still need to keep the clone here. 
`self.prune_object_hash(hash)` borrows self as mut, so I cant borrow self as immutable which is what happens because I want to borrow the hash in the vec.